packages/powersync_attachments_stream/lib/src/local_storage.dart

```diff
@@ -0,0 +1,84 @@
+import 'dart:convert';
+import 'dart:io';
+import 'package:path/path.dart' as p;
+
+abstract class LocalStorage {
+  /// Saves binary data to storage with an identifier
+  /// 
+  /// [id] - Unique identifier for the stored data
+  /// [bytes] - Binary data to store
+  /// [mediaType] - Optional MIME type of the data (e.g., 'image/jpeg')
+  /// [metadata] - Optional key-value pairs for additional data information
+  Future<void> save(String id, List<int> bytes, {String? mediaType, Map<String, dynamic>? metadata});
+  
+  /// Retrieves binary data by identifier
+  /// 
+  /// [id] - Unique identifier of the data to retrieve
+  /// Returns the binary data if found, null otherwise
+  Future<List<int>?> read(String id);
+  
+  /// Removes data and its metadata from storage
+  /// 
+  /// [id] - Unique identifier of the data to delete
+  Future<void> delete(String id);
+  
+  /// Retrieves metadata associated with stored data
+  /// 
+  /// [id] - Unique identifier of the data
+  /// Returns metadata map if found, null otherwise
+  Future<Map<String, dynamic>?> getMetadata(String id);
+}
+
+class IOLocalStorage implements LocalStorage {
+  final Directory baseDir;
+
+  IOLocalStorage(this.baseDir);
+
+  File _fileFor(String id) => File(p.join(baseDir.path, id));
+  File _metaFileFor(String id) => File(p.join(baseDir.path, '$id.meta.json'));
+
+  @override
+  Future<void> save(String id, List<int> bytes, {String? mediaType, Map<String, dynamic>? metadata}) async {
+    await baseDir.create(recursive: true);
+    await _fileFor(id).writeAsBytes(bytes);
+    if (mediaType != null || metadata != null) {
```

Comment:
I'm not sure about saving the media type like this, what's the motivation for not storing it in the attachments table instead?